crates/oci-distribution/src/reference.rs

```diff
@@ -0,0 +1,82 @@
+use std::convert::{Into, TryFrom};
+
+// currently, the library only accepts modules tagged in the following structure:
+// <registry>/<repository>:<tag>
+// for example: webassembly.azurecr.io/hello:v1
+#[derive(Clone)]
+pub struct Reference {
+    whole: String,
+    slash: usize,
+    colon: usize,
+}
+
+impl Reference {
+    pub fn whole(&self) -> &str {
+        &self.whole
+    }
+
+    pub fn registry(&self) -> &str {
+        &self.whole[..self.slash]
+    }
+
+    pub fn repository(&self) -> &str {
+        &self.whole[self.slash + 1..self.colon]
+    }
+
+    pub fn tag(&self) -> &str {
+        &self.whole[self.colon + 1..]
+    }
+
+    pub fn to_v2_manifest_url(&self) -> String {
+        format!(
+            "https://{}/v2/{}/manifests/{}",
```

Comment:
Shouldn't this be using the `OCI_VERSION_2` constant? We would need the const to be `pub(crate)` but I think we should probably use it here

Reply:
What would be the reasoning for doing string interpolation to insert a constant? That just increases the runtime burden.

Reply:
As a side note, we might also want to support HTTP-only registries.
(Had to manually change this to HTTP to use a local registry.)

Reply:
My thinking was more along the DRY route here. But I don't have strong feelings one way or another here.

Reply:
Oh! You can do http-only? The spec made it sound like it had to be https!

Reply:
DRY would make sense if this was something that could ever change, or if I were removing code paths. But it's not doing either.

The V2 version of the spec cannot switch to different URL formats.

By "drying" here, you'd actually be introducing new code paths. And you get harder-to-read code, longer code, and a small performance delay. So I would argue that DRY would be misapplied in this context.